cpp/open3d/pipelines/registration/Registration.cpp

```diff
@@ -194,152 +192,160 @@ RegistrationResult RegistrationRANSACBasedOnCorrespondence(
-                                             transformation)) {
-                        check = false;
-                        break;
-                    }
-                }
-                if (!check) continue;
-                transformation = estimation.ComputeTransformation(
-                        source, target, ransac_corres);
-                check = true;
-                for (const auto &checker : checkers) {
-                    if (checker.get().require_pointcloud_alignment_ &&
-                        !checker.get().Check(source, target, ransac_corres,
+                    if (!checker.get().Check(source, target, ransac_corres,
                                              transformation)) {
                         check = false;
                         break;
                     }
                 }
                 if (!check) continue;
+
                 geometry::PointCloud pcd = source;
                 pcd.Transform(transformation);
-                auto this_result = GetRegistrationResultAndCorrespondences(
-                        pcd, target, kdtree, max_correspondence_distance,
+                auto result = EvaluateRANSACBasedOnCorrespondence(
+                        pcd, target, corres, max_correspondence_distance,
                         transformation);
-                if (this_result.fitness_ > result_private.fitness_ ||
-                    (this_result.fitness_ == result_private.fitness_ &&
-                     this_result.inlier_rmse_ < result_private.inlier_rmse_)) {
-                    result_private = this_result;
-                }
-#pragma omp critical
-                {
-                    total_validation = total_validation + 1;
-                    if (total_validation >= criteria.max_validation_)
-                        finished_validation = true;
+
+                if (result.IsBetterRANSACThan(best_result_local)) {
+                    best_result_local = result;
+
+                    // Update exit condition if necessary
+                    double exit_itr_d =
+                            std::log(1.0 - criteria.confidence_) /
+                            std::log(1.0 - std::pow(result.fitness_, ransac_n));
+                    exit_itr_local =
+                            exit_itr_d < double(criteria.max_iteration_)
+                                    ? static_cast<int>(std::ceil(exit_itr_d))
+                                    : exit_itr_local;
                 }
-            }  // end of if statement
-        }      // end of for-loop
+            }  // if < exit_itr_local
+        }      // for loop
 #pragma omp critical
         {
-            if (result_private.fitness_ > result.fitness_ ||
-                (result_private.fitness_ == result.fitness_ &&
-                 result_private.inlier_rmse_ < result.inlier_rmse_)) {
-                result = result_private;
+            if (best_result_local.IsBetterRANSACThan(best_result)) {
```

Comment:
In relation to the comment above, I'm not sure I get the value of identifying the best result thread-locally first, and then reducing it globally, in particular because the reduction here a rather check and operation, so you might just end up having idle threads, that could instead work on other samples in the loop above.

Reply:
See below